### PR TITLE
Feature: Improve robustness and capabilities of variable inlining

### PR DESCRIPTION
**Pull request by jbrette**
This PR is addressing the following issue: https://github.com/kubernetes-sigs/kustomize/issues/1190

It allows add the ability to inline/copy a tree first from one K8s object to another and let the standard kustomize transformation proceed.

This is really useful because it allows the reuse of complex tree structure (for instance PodTemplate) accross K8s object and does not force the user to create one variable per field.

### CODE REVIEW

**Comment by jbrette** [comment]
gentle quick to retriger build

File: pkg/transformers/refvars.go [hunk 1]
```diff
@@ -41,49 +43,84 @@ func NewRefVarTransformer(
 	}
 }
 
+// replacePrimitiveType checks if the field is a string. If not it will return
+// the field. In case it is a string, it will expand the field, and perform
+// a deepCopy of the expanded value in case it is not of primitiv type.
```
Comment (monopole) [inline pkg/transformers/refvars.go]: // replaceVarsIfStringField checks if the incoming field contains a string.
// If so, the field is processed to replace variables.
// If not, the field is returned as is.

This names the function like a trailing conditional, e.g. http://jerodsanto.net/2015/03/trailing-conditionals-considered-harmful/

File: pkg/transformers/refvars.go [hunk 2]
```diff
@@ -41,49 +43,84 @@ func NewRefVarTransformer(
...
+func (rv *RefVarTransformer) replacePrimitiveType(a interface{}) interface{} {
+	s, ok := a.(string)
+	if !ok {
+		// This field is not of string type.
+		// It can not contain a $(VAR)
```
Comment (monopole) [inline pkg/transformers/refvars.go]: s/can not/cannot/

File: pkg/transformers/refvars.go [hunk 3]
```diff
@@ -41,49 +43,84 @@ func NewRefVarTransformer(
...
+		return a
+	}
+
+	// This field can potientially contain a $(VAR)
```
Comment (monopole) [inline pkg/transformers/refvars.go]: // This field may contain a $(VAR).

File: pkg/transformers/refvars_helper.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	case []interface{}:
+		return castAndMergeSlice(dstVal, srcVal)
+	case map[string]interface{}:
+		return castAndMergeMap(dstVal, srcVal)
+	default:
+		switch dstValType := dstVal.(type) {
+		case nil:
+			return srcVal, nil
+		default:
+			if dstValType != srcValType {
```
Comment (monopole) [inline pkg/transformers/refvars_helper.go]: this if doesn't matter

File: pkg/transformers/refvars.go [hunk 5]
```diff
@@ -41,49 +43,84 @@ func NewRefVarTransformer(
...
-		// of string type.
-		return expansion.Expand(s, rv.mappingFunc), nil
+	case string:
+		// Attempt to expand this simple field
+		return rv.replacePrimitiveType(in), nil
 	case nil:
 		return nil, nil
 	default:
-		return "", fmt.Errorf("invalid type encountered %T", vt)
+		// This field not contain a $(VAR) since it is not of string type.
```
Comment (monopole) [inline pkg/transformers/refvars.go]: s/not/cannot/

File: pkg/transformers/refvars.go [hunk 6]
```diff
@@ -41,49 +43,84 @@ func NewRefVarTransformer(
...
+// replaceParentInline allows to inline the complex tree of a variable
+// at the same time it allows to replace and patch individual member of
+// that inlined tree.
+func (rv *RefVarTransformer) replaceParentInline(inMap map[string]interface{}) (interface{}, error) {
+	s, _ := inMap[parentInline].(string)
+
+	inlineValue := expansion.Expand(s, rv.mappingFunc)
+	newMap, ok := inlineValue.(map[string]interface{})
+	if !ok {
+		log.Printf("inlining issue with %s", inlineValue)
```
Comment (monopole) [inline pkg/transformers/refvars.go]: Traditionally, kustomize doesn't log errors and continue, opting to instead just fail the build with an error.

To the end use, logged errors and still emitting output just raises questions like "is this right?"

File: pkg/transformers/refvars.go [hunk 7]
```diff
@@ -41,49 +43,84 @@ func NewRefVarTransformer(
...
+	expandedValue := expansion.Expand(s, rv.mappingFunc)
+
+	// Let's perform a deep copy if we didn't inline
+	// a primitive type
+	return deepCopy(expandedValue)
+}
+
+// replaceParentInline allows to inline the complex tree of a variable
+// at the same time it allows to replace and patch individual member of
+// that inlined tree.
```
Comment (monopole) [inline pkg/transformers/refvars.go]: Is this the right use of the word "parent"?  I don't see a parent/child relation here.  Not sure it helps in the function name.

File: pkg/transformers/refvars_helper.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+			}
+			dstMap[key] = mergedVal
+		} else {
+			dstMap[key] = srcVal
+		}
+	}
+	return dstMap, nil
+}
+
+// TODO(jeb): Investigate the usage of obj->json->obj conversion
```
Comment (ian-howell) [inline pkg/transformers/refvars_helper.go]: @monopole Do you have any opinion on using a json patch strategy here? Maybe something like what's being done in the [k8sdeps/transformer/patch package](https://github.com/kubernetes-sigs/kustomize/blob/master/k8sdeps/transformer/patch/transformer.go#L66). I think we could do away with a lot of the code in this file if we switched to something similar.

File: pkg/transformers/refvars.go (hunk 1 above)
Comment (jbrette) [reply pkg/transformers/refvars.go]: Thank you for the advise. Renamed method to replaceStringField

File: pkg/transformers/refvars.go (hunk 2 above)
Comment (jbrette) [reply pkg/transformers/refvars.go]: done

File: pkg/transformers/refvars.go (hunk 3 above)
Comment (jbrette) [reply pkg/transformers/refvars.go]: done

File: pkg/transformers/refvars.go (hunk 6 above)
Comment (jbrette) [reply pkg/transformers/refvars.go]: changed to be aligned with kustomize traditional behavior. Removed the log statement and returned an error

File: pkg/transformers/refvars.go (hunk 5 above)
Comment (jbrette) [reply pkg/transformers/refvars.go]: done

File: pkg/transformers/refvars_helper.go (hunk 4 above)
Comment (jbrette) [reply pkg/transformers/refvars_helper.go]: thanks. fixed

File: pkg/transformers/refvars.go (hunk 7 above)
Comment (jbrette) [reply pkg/transformers/refvars.go]: rename into inlineIntoParentNode and added comments.

**Comment by jbrette** [comment]
@monopole We were really happy to be able to use kustomize as an improvement over one of the component of our product architecture. To give more better understanding of what we are trying to achieve, we joined three examples of the usage of that feature. In each case, in order to respect the yaml syntax, we could not do:

```
construct:
   parent-node: $(var-to-inline)
   child-field: child-value
```

```
construct:
   parent-node: 
      parent-inline: $(var-to-inline)
      child-field: child-value
``` 

Here are the examples:
1) Check out: https://github.com/keleustes/airship-treasuremap/tree/inline and in particular:
https://github.com/keleustes/airship-treasuremap/blob/inline/global/config/endpoints.yaml#L132 which should be replaced by: https://github.com/keleustes/airship-treasuremap/blob/inline/global/config/service_accounts.yaml#L227

2) More simple version of the same concept:
https://github.com/keleustes/kustomize/blob/allinone/examples/allinone/base/wordpress/deployment.yaml#L15

3) The example we added in the pkg/target:
https://github.com/keleustes/kustomize/blob/inline/pkg/target/variableref_test.go#L130

**Comment by monopole** [comment]
hey folks, this looks good, just want to circulate a bit and ponder, and consider the longer term implications on the api.

**Comment by jbrette** [comment]
@monopole Latest series of forced push where simply to validate everything was still working ok.

**Comment by jbrette** [comment]
@monopole Add another test demonstrating composition using enhanced inline feature,

**Comment by rdubya16** [comment]
I was wondering if this PR would help with the following use case? We have a bunch of cronjobs with mostly boilerplate configuration with the only thing that varies are the container args, name, and schedule. We would like to not have to have this yaml duplicated over and over and be able to parameterize only what is needed in these jobs.

**Comment by jbrette** [comment]
@rdubya16 We had the same question from another use. Have a look at:
- [Issue](https://github.com/kubernetes-sigs/kustomize/issues/1339)
- [Example](https://github.com/keleustes/kustomize/tree/allinone/examples/issues/issue_1339)
- [Another recent example](https://github.com/keleustes/kustomize/tree/allinone/examples/issues/issue_1251_f)

**Comment by tkellen** [comment]
@monopole / @Liujingfang1 Apologies for the repeated pings over the last few months but, is there any chance either of you will be giving this or #1217 a review before the end of the year?

**Comment by jbrette** [comment]
Seems that PR simplifying the usage of kustomize can't merge. Closing it.

**Pull request closed without merging**